### PR TITLE
Implement texture functions for Metal target

### PR DESCRIPTION
**Pull request by jkwak-work**
This commit is to implement texture functions for Metal target.

The following functions are implemented and tested.
 - GetDimensions()
 - CalculateLevelOfDetail()
 - CalculateLevelOfDetailUnclamped()
 - Sample()
 - SampleBias()
 - SampleLevel()
 - SampleCmp()
 - SampleCmpLevelZero()
 - Gather()
 - SampleGrad()
 - Load()

Metal has limited support for the texture functions compared to HLSL.
 - LOD is not supported for 1D texture,
 - Depth textures are limited to 2D, 2DArray, Cube and CubeArray textures.
 - "Offset" variants are limited to 2D, 2DArray, 2D-Depth, 2DArray-Depth and 3D textures.

The functions that cannot be implemented for Metal should properly be handled by the capability system later.

### CODE REVIEW

File: tests/metal/texture.slang [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+
+        // METAL: d2D{{.*}}.sample_compare(
+        // METALLIB: call {{.*}}.sample_compare_depth_2d.f32(
+        && float(1) == d2D.SampleCmp(shadowSampler, float2(u, u), 0)
+
+#if defined(TEST_AFTER_FIXING_CAPABILITY_PROBLEM)
+        // M-ETAL: d2DArray{{.*}}.sample_compare(
+        // M-ETALLIB: call {{.*}}.sample_compare_depth_2d_array.f32(
+        && float(1) == d2DArray.SampleCmp(shadowSampler, normalize(float3(u, 1 - u, u)), 0)
+#endif
```
Comment (jkwak-work) [inline tests/metal/texture.slang:488]: @ArielG-NV ,can you take a look at these two cases and give me some help on why the second one doesn't work with the capability system?
Both of them use the same signature,
```
    [__readNone]
    [ForceInline]
    [require(glsl_hlsl_metal_spirv, texture_shadowlod)]
    float SampleCmp(SamplerComparisonState s, vector<float, Shape.dimensions+isArray> location, float compareValue)
```
How come one of them works and the other one doesn't?
When the issue occurs, the exact error message is following,
```
tests/metal/texture.slang(661): error 36107: entrypoint 'computeMain' requires capability 'hlsl + fragment | hlsl + compute | hlsl + raygen | hlsl + intersection | hlsl + anyhit | hlsl + closesthit | hlsl + miss | hlsl + callable | glsl | spirv_1_0', which is incompatible with the current compilation target 'metal + compute'.
void computeMain()
     ^~~~~~~~~~~
}
```

Comment (ArielG-NV) [reply tests/metal/texture.slang:488]: debugging what is wrong locally, I will update when I figure out what went wrong. 
There were a few issues with the old capability system skipping to check certain capabilities that I found while moving over to the new system, I will check if this is one of these scenarios or if something else happened.

**Review by csyonghe — COMMENTED** [review]
We should also make sure `lowerCombinedTextureSamplers` is called for metal, and combined texture samplers works.

File: tests/metal/texture.slang (hunk 1 above)
Comment (ArielG-NV) [reply tests/metal/texture.slang:488]: I currently only see the error:
```
standard error = {
tests/metal/texture.slang(142): note 99999: an internal error threw an exception while working on code near this location
    t1D.GetDimensions(width);
                     ^
(0): error 99999: Slang compilation aborted due to an exception of class Slang::InternalError: assert failure: (0 <= argIndex) && (argIndex < m_argCount)
}
```
is your local code pushed to your branch (so I may test it)?

Comment (ArielG-NV) [reply tests/metal/texture.slang:488]: update: everything works after latest PR

File: source/slang/slang-stdlib-textures.cpp [hunk 2]
```diff
@@ -184,27 +192,56 @@ void TextureTypeInfo::writeGetDimensionFunctions()
...
+
+            StringBuilder metal;
+            const char* metalMipLevel = "0";
+
             if (includeMipInfo)
+            {
+                ++paramCount;
                 params << "uint mipLevel, ";
 
+                if (baseShape != SLANG_TEXTURE_1D);
```
Comment (jkwak-work) [inline source/slang/slang-stdlib-textures.cpp]: Note: there was a bug that is fixed in the latest PR, @ArielG-NV

**Comment by jkwak-work** [comment]
> We should also make sure `lowerCombinedTextureSamplers` is called for metal, and combined texture samplers works.

I see. This function looks to be creating a default sampler for the combined textures.
I will use it to make the combined textures working.

**Comment by jkwak-work** [comment]
> > We should also make sure `lowerCombinedTextureSamplers` is called for metal, and combined texture samplers works.
> 
> I see. This function looks to be creating a default sampler for the combined textures. I will use it to make the combined textures working.

I got the combined textures working.
But I will prepare a separate PR for it because this change is already big.

File: source/slang/hlsl.meta.slang [hunk 3]
```diff
@@ -872,14 +872,16 @@ extension __TextureImpl<T,Shape,isArray,isMS,sampleCount,access,isShadow,0,forma
...
     float CalculateLevelOfDetail(SamplerState s, TextureCoord location)
     {
         __requireComputeDerivative();
         __target_switch
         {
         case hlsl:
-            __intrinsic_asm "CalculateLevelOfDetail";
+            __intrinsic_asm ".CalculateLevelOfDetail";
+        case metal:
+            __intrinsic_asm "$0.calculate_clamped_lod($1, $2)";
```
Comment (csyonghe) [inline source/slang/hlsl.meta.slang:884]: Can be simplified to ".calculate_clamp_lod" so it is consistent with the HLSL case.

File: source/slang/hlsl.meta.slang [hunk 4]
```diff
@@ -1520,13 +1959,32 @@ vector<TElement,4> __glsl_gatherCmp(__TextureImpl<T, Shape, isArray, 0, sampleCo
...
+                }
+                else
+                {
+                    // Tv gather_compare(sampler s, float2 coord, float compare_value, int2 offset = int2(0)) const
+                    __intrinsic_asm "$0.gather_compare($1, $2, $3, $4)";
+                }
+            }
+        }
+        // TODO: This needs to be handled by the capability system
+        __intrinsic_asm "<Metal support gather_compare with offset only for 2D depth>";
```
Comment (csyonghe) [inline source/slang/hlsl.meta.slang:1987]: The capability system won't have the feature to detect things in a branch like this. Perhaps we can consider a __intrinsic_asm "<illegal: message>"; syntax and then we issue a diagnostic with source loc pointing to the call site.

**Review by csyonghe — APPROVED** [review]

File: source/slang/hlsl.meta.slang (hunk 3 above)
Comment (jkwak-work) [reply source/slang/hlsl.meta.slang:884]: Good catch. I will make the change and check if there are more like this

File: source/slang/hlsl.meta.slang (hunk 4 above)
Comment (jkwak-work) [reply source/slang/hlsl.meta.slang:1987]: Is it possible to issue a diagnostic from hlsl.meta.slang? Or should it be implemented from c++ side?

File: source/slang/hlsl.meta.slang [hunk 5]
```diff
@@ -1382,13 +1690,77 @@ Array<T,4> __makeArray<T>(T v0, T v1, T v2, T v3);
...
+            {
+                if (isShadow == 1)
+                {
+                    // Tv gather(sampler s, float2 coord, uint array, int2 offset = int2(0)) const
+                    __intrinsic_asm "$0.gather($1, ($2).xy, as_type<uint>(($2).z))";
+                }
+                else
+                {
+                    // Tv gather(sampler s, float2 coord, uint array, int2 offset = int2(0), component c = component::x) const
+                    __intrinsic_asm "$0.gather($1, ($2).xy, as_type<uint>(($2).z), int2(0), metal::component($3))";
```
Comment (jkwak-work) [inline source/slang/hlsl.meta.slang:1713]: I think I made a mistake here. `as_type` is `reinterprete_cast` and I intended to use it as `static_cast`. I was having some troubles with type casting in Metal and only thing worked was `as_type` when I tried it. I will fix this tomorrow.

File: tests/bindings/multi-file-shared.slang [hunk 6]
```diff
@@ -7,6 +7,10 @@ PUBLIC float4 use(float  val) { return val; };
 PUBLIC float4 use(float2 val) { return float4(val,0.0,0.0); };
 PUBLIC float4 use(float3 val) { return float4(val,0.0); };
 PUBLIC float4 use(float4 val) { return val; };
+
+#ifdef __SLANG__
+[require(hlsl)]
+#endif
```
Comment (jkwak-work) [inline tests/bindings/multi-file-shared.slang:13]: I am not sure why I have to have this one here. Without this change, this test was failing due to the capability system warning. I suspect that there might be a bug in the capability checking logic.

**Comment by csyonghe** [comment]
@jkwak-work Is this ready for merge?

**Comment by jkwak-work** [comment]
Yes it is ready for review.
I wanted to sort out the capability side issue, but it can be done after the merge

File: source/slang/hlsl.meta.slang (hunk 4 above)
Comment (csyonghe) [reply source/slang/hlsl.meta.slang]: There isn't such a construct yet. We could implement something like `kIROp_CodeGenError` inst that will result in a code-gen time error if it doesn't get dead code eliminated at emit time.

File: tests/bindings/multi-file-shared.slang (hunk 6 above)
Comment (csyonghe) [reply tests/bindings/multi-file-shared.slang:13]: That seems problematic.
Have you tried this with Ariel's pending PR?

**Review by csyonghe — DISMISSED** [review]

File: tests/bindings/multi-file-shared.slang (hunk 6 above)
Comment (csyonghe) [reply tests/bindings/multi-file-shared.slang:13]: If this test was not failing before, it implies something is broken, and we should probably fix it before merging.

If Ariel's pending PR fixes this issue, then let's merge this PR after his.

**Comment by jkwak-work** [comment]
@ArielG-NV helped me to find that the use of `normalize` was causing one of two issues.
The latest change has the fix.

I am still investigating why the change in `tests/binding/multi-file-shared.slang` is needed.

File: tests/bindings/multi-file-shared.slang (hunk 6 above)
Comment (jkwak-work) [reply tests/bindings/multi-file-shared.slang:13]: I figured why this was causing a trouble.
I had to add `metal` to the following aliases,
```
alias any_target = hlsl | glsl | c | cpp | cuda | spirv;
alias any_textual_target = hlsl | glsl | c | cpp | cuda;
alias any_gfx_target = hlsl | glsl | spirv;
```
I am going to submit my fix shortly.

File: tests/metal/texture.slang [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+    // Metal doesn't support some features, and we need a new test that
+    // can check if the capability system can error them out properly.
+    #define NEED_TO_TEST_FOR_METAL_CAPABILITY 1
+#endif
+
+#if defined(EMIT_SOURCE)
+    // It appears that Slang-test doesn't initialize the depth cube texture
+    // properly.
+    #define TEST_WHEN_DEPTH_CUBE_WORKS
+#endif
```
Comment (jkwak-work) [inline tests/metal/texture.slang:19]: I will create a new issue for this.
The problem is that when the type is depth-cube texture, our "TEST_INPUT" method is not able to set it up properly.

File: source/slang/hlsl.meta.slang (hunk 4 above)
Comment (jkwak-work) [reply source/slang/hlsl.meta.slang]: I will create a new issue for this.

**Comment by jkwak-work** [comment]
All issues are addressed properly.
The change is ready for review and merge.

**Review by csyonghe — APPROVED** [review]

**Pull request merged**